p2p/net/swarm/swarm_dial.go

```diff
@@ -465,10 +465,21 @@ func (s *Swarm) dialAddr(ctx context.Context, p peer.ID, addr ma.Multiaddr) (con
 	return connC, nil
 }
 
+var ConnSetupTimeout = time.Minute * 5
+
 // dialConnSetup is the setup logic for a connection from the dial side. it
 // needs to add the Conn to the StreamSwarm, then run newConnSetup
 func dialConnSetup(ctx context.Context, s *Swarm, connC conn.Conn) (*Conn, error) {
 
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		deadline = time.Now().Add(ConnSetupTimeout)
+	}
+
+	if err := connC.SetDeadline(deadline); err != nil {
```

Comment:
Got it, you use the SetDeadline like a timeout :)

re: adding SetDeadline for muxed streams - A muxed stream should not timeout, since a conn was already established, so do we really need to make it part of the API?